src/Evolution/DiscontinuousGalerkin/SlopeLimiters/HwenoModifiedSolution.cpp

```diff
@@ -0,0 +1,340 @@
+
+  // Cache will store quantities only for directions that have neighbors
+  const std::vector<Direction<VolumeDim>>
+      directions_with_neighbors = [&element]() noexcept {
+    std::vector<Direction<VolumeDim>> result;
+    for (const auto& dir_and_neighbors : element.neighbors()) {
+      result.push_back(dir_and_neighbors.first);
+    }
+    return result;
+  }
+  ();
+
+  for (const auto& dir : directions_with_neighbors) {
+    interpolation_matrices[dir] =
+        volume_interpolation_matrix(mesh, mesh, element, dir);
+    quadrature_weights_dot_interpolation_matrices[dir] = apply_matrices(
+        std::array<Matrix, 1>{{trans(interpolation_matrices.at(dir))}},
+        quadrature_weights, Index<1>(quadrature_weights.size()));
+  }
+
+  for (const auto& primary_dir : directions_with_neighbors) {
+    if (directions_with_neighbors.size() == 1) {
+      // With a single neighbor, there can be no neighbors to exclude.
+      std::vector<Direction<VolumeDim>> nothing_to_exclude{{}};
+      // To reuse the same data structure from the more common `else` branch,
+      // here we stick the data into the (normally nonsensical) slot where
+      // `dir_to_exclude == primary_dir`.
+      a_inverse_matrices[primary_dir][primary_dir] = a_inverse_matrix(
+          element, mesh, quadrature_weights, interpolation_matrices,
+          quadrature_weights_dot_interpolation_matrices, primary_dir,
+          nothing_to_exclude);
+    } else {
+      // With multiple neighbors, there can be 1+ neighbors to exclude. Because
+      // the case where 2+ secondary neighbors must be excluded is rare, we do
+      // not cache it. So we only handle the case of 1 neighbor to exclude.
+      for (const auto& dir_to_exclude : directions_with_neighbors) {
+        // Skip the nonsensical case where the primary and excluded neighbors
+        // are the same.
+        if (dir_to_exclude == primary_dir) {
+          continue;
+        }
+        a_inverse_matrices[primary_dir][dir_to_exclude] = a_inverse_matrix(
+            element, mesh, quadrature_weights, interpolation_matrices,
+            quadrature_weights_dot_interpolation_matrices, primary_dir,
+            {{dir_to_exclude}});
+      }
+    }
+  }
+}
+
+template <size_t VolumeDim>
+const Matrix& ConstrainedFitCache<VolumeDim>::retrieve_a_inverse_matrix(
+    const Direction<VolumeDim>& primary_direction,
+    const std::vector<Direction<VolumeDim>>& directions_to_exclude) const
+    noexcept {
+  if (LIKELY(directions_to_exclude.size() == 1)) {
+    // The typical case
+    return a_inverse_matrices.at(primary_direction)
+        .at(directions_to_exclude[0]);
+  } else if (directions_to_exclude.empty()) {
+    // The atypical case is cached in a normally nonsensical slot to reuse the
+    // data structure.
+    return a_inverse_matrices.at(primary_direction).at(primary_direction);
+  } else {
+    ERROR("Asking for an uncached matrix, with multiple excluded neighbors");
```

Comment:
I'd add a (very brief) justification in the error message because if I got this error my reaction would be "Well then compute if it's not cached instead of giving me an error... Francois plz" :)